### PR TITLE
[CORE] Add InputIteratorTransformer to decouple ReadRel and iterator index

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

This pr adds a new transformer `InputIteratorTransformer` to connect SparkPlan and TransformSupport. It replace the original `ColumnarInputAdapter` to provide the substrait plan ReadRel for the child columnar iterator, so that the TransformSupport always has input. It would be transformed to ValueStreamNode at native Velox side.

The reason is that:
1. decopule with scan transofrmer, the scan transofrmer does not need iterator index
2. make the whole stage transform framework more clear, all the TransformSupport does not need to check if the child is a TransformSupport and build a ReadRel by theirself
3. make the metrics framework more clear, the metrics should belong to the input iterator rather than the operators
4. it aligns with native side, e.g., velox backend will transform it to ValueStreamNode

## How was this patch tested?

Pass CI and manually test

image

### CODE REVIEW

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala [hunk 1]
```diff
@@ -37,6 +37,20 @@ class MetricsApiImpl extends MetricsApi with Logging {
...
   }
 
+  override def genInputIteratorTransformerMetrics(
+      sparkContext: SparkContext): Map[String, SQLMetric] = {
+    Map(
+      "cpuCount" -> SQLMetrics.createMetric(sparkContext, "cpu wall time count"),
+      "wallNanos" -> SQLMetrics.createNanoTimingMetric(sparkContext, "totaltime of batch scan"),
+      "outputVectors" -> SQLMetrics.createMetric(sparkContext, "number of output vectors")
+    )
+  }
```
Comment (Yohahaha) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:47]: could you also add `inputVectors` ? seems shuffle read does not show that.

Comment (ulysses-you) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:47]: Does not ValueStream inputVectors always same with outputVectors ?

Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:47]: you are right.

**Comment by ulysses-you** [comment]
Run Gluten Clickhouse CI

**Comment by ulysses-you** [comment]
cc @Yohahaha @zhztheplayer @zhouyuan @zzcclp thank you

**Comment by zhztheplayer** [comment]
cc @rui-mo

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala [hunk 2]
```diff
@@ -37,6 +37,20 @@ class MetricsApiImpl extends MetricsApi with Logging {
     MetricsUtil.updateNativeMetrics(child, relMap, joinParamsMap, aggParamsMap)
   }
 
+  override def genInputIteratorTransformerMetrics(
+      sparkContext: SparkContext): Map[String, SQLMetric] = {
+    Map(
+      "cpuCount" -> SQLMetrics.createMetric(sparkContext, "cpu wall time count"),
+      "wallNanos" -> SQLMetrics.createNanoTimingMetric(sparkContext, "totaltime of batch scan"),
```
Comment (zhztheplayer) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:44]: Is it accurate to use "totaltime of batch scan"? I thought the transformer is common for both scan and other inputs.

File: gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseTransformStages.scala [hunk 3]
```diff
@@ -19,13 +19,58 @@ package org.apache.spark.sql.execution
...
 import java.util.concurrent.atomic.AtomicInteger
 
+import scala.collection.JavaConverters._
+
+/**
+ * A bridge to connect [[SparkPlan]] and [[TransformSupport]] and provide the substrait plan
+ * `ReadRel` for the child columnar iterator, so that the [[TransformSupport]] always has input. It
+ * would be transformed to `ValueStreamNode` at native side.
+ */
+case class InputIteratorTransformer(child: SparkPlan) extends UnaryTransformSupport {
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseTransformStages.scala:43]: Should we combine `InputIteratorTransformer` and `ColumnarInputAdapter` later on?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala (hunk 2 above)
Comment (ulysses-you) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/MetricsApiImpl.scala:44]: good catch, I think it is copied from a wrong code path..

File: gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseTransformStages.scala (hunk 3 above)
Comment (ulysses-you) [reply gluten-core/src/main/scala/org/apache/spark/sql/execution/ColumnarCollapseTransformStages.scala:41]: It seems we do not need `ColumnarInputAdapter` any more

**Comment by Yohahaha** [comment]
LGTM! thank you for this great patch.

**Comment by zzcclp** [comment]
Good patch. Sorry for the late review, if there is any issue for ch backend, I will fix later.

**Pull request merged**